src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/JsonDictionaryConverter.cs

```diff
@@ -83,6 +83,7 @@ protected static JsonConverter<T> GetConverter<T>(JsonTypeInfo typeInfo)
             ref ReadStack state,
             [MaybeNullWhen(false)] out TDictionary value)
         {
+            JsonTypeInfo keyTypeInfo = state.Current.JsonTypeInfo.KeyTypeInfo!;
```

Comment:
Ensures any unsupported key type errors are forced at the start of serialization.